phoenix-core/src/test/java/org/apache/phoenix/query/SplitKeyRangesByBoundariesTest.java

```diff
@@ -0,0 +1,214 @@
+
+        // 1, region boundaries is null, return null
+        testCases.addAll(Arrays.asList(new Object[][] { { expected(null), boundaries(null),
+                keyRanges(PInteger.INSTANCE
+                        .getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(2), false)) } }));
+
+        // 2, key ranges is null, return null
+        testCases.addAll(Arrays.asList(new Object[][] {
+                { expected(null), boundaries(Bytes.toBytes(2)), keyRanges(null) } }));
+
+        // 3, empty key ranges, return null
+        testCases.addAll(Arrays.asList(new Object[][] {
+                { expected(null), boundaries(Bytes.toBytes(2)), keyRanges() } }));
+
+        // 4, single regions, single key range
+        testCases.addAll(Arrays.asList(new Object[][] {
+                {
+                        expected(new KeyRange[][] {
+                                { PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(5), false), }, }),
+                        boundaries(),
+                        keyRanges(PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(5), false))
+                } }));
+
+        // 5, single regions, multiple key ranges
+        testCases.addAll(Arrays.asList(new Object[][] {
+                {
+                        expected(new KeyRange[][] {
+                                {
+                                        PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(2), false),
+                                        PInteger.INSTANCE.getKeyRange(Bytes.toBytes(3), true, Bytes.toBytes(5), false),
+                                },
+                        }),
+                        boundaries(),
+                        keyRanges(
+                                PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(2), false),
+                                PInteger.INSTANCE.getKeyRange(Bytes.toBytes(3), true, Bytes.toBytes(5), false)
+                        )
+                } }));
+
+        // 6, Multiple regions, one key range covers all regions.
+        testCases.addAll(Arrays.asList(new Object[][] {
+                {
+                        expected(new KeyRange[][] {
+                            { PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(2), false), },
+                            { PInteger.INSTANCE.getKeyRange(Bytes.toBytes(2), true, Bytes.toBytes(4), false), },
+                            { PInteger.INSTANCE.getKeyRange(Bytes.toBytes(4), true, Bytes.toBytes(5), false), }, }),
+                        boundaries(Bytes.toBytes(2), Bytes.toBytes(4)),
+                        keyRanges(PInteger.INSTANCE.getKeyRange(Bytes.toBytes(1), true, Bytes.toBytes(5), false))
+                } }));
+
+        // 7, Multiple regions, one key range is in the first region.
+        testCases.addAll(Arrays.asList(new Object[][] {
+                {
+                        expected(new KeyRange[][] {
```

Comment:
Shouldn't we have trailing empty arrays?  In test case 8 we have one leading array but shouldn't we also check the trailing ones exist?

Reply:
I didn't see anything in the transform function ARRAY_TO_LIST that i think is handling this?

Reply:
I changed to return List<Pair<Integer, List>>, with latest change, this concern should have gone.